### PR TITLE
 Prefer fetching small PoVs from backing group

### PR DESCRIPTION
**Pull request by sandreim**
Fixes https://github.com/paritytech/polkadot/issues/6832

### CODE REVIEW

File: node/network/availability-recovery/src/lib.rs [hunk 1]
```diff
@@ -99,15 +99,20 @@ const TIMEOUT_START_NEW_REQUESTS: Duration = CHUNK_REQUEST_TIMEOUT;
...
 	/// Do not request data from the availability store.
 	/// This is the useful for nodes where the
 	/// availability-store subsystem is not expected to run,
 	/// such as collators.
 	bypass_availability_store: bool,
-
+	/// If true, we first try backers.
 	fast_path: bool,
+	/// Large PoV threshold. Only used when `fast_path` is `false.
+	small_pov_limit: Option<usize>,
```
Comment (eskimor) [inline node/network/availability-recovery/src/lib.rs:115]: Probably better to unify this with `fast_path` and make it an enum like:

```rust
enum RecoveryStrategy {
  BackersFirst,
  BackersFirstIfSizeBelow(usize),
  AlwaysFullRecovery,
}
```

File: node/network/availability-recovery/src/lib.rs [hunk 2]
```diff
@@ -880,6 +886,29 @@ async fn launch_recovery_task<Context>(
 		bypass_availability_store,
 	};
 
+	if let Some(small_pov_limit) = small_pov_limit {
+		// Get our own chunk size to get an estimate of the PoV size.
+		let chunk_size: Result<Option<usize>, error::Error> =
+			query_chunk_size(ctx, candidate_hash).await;
+		if let Ok(Some(chunk_size)) = chunk_size {
+			let pov_size_estimate = chunk_size.saturating_mul(session_info.validators.len() / 3);
```
Comment (eskimor) [inline node/network/availability-recovery/src/lib.rs:894]: Better do the division on the outer level - should work better with smaller networks.

File: node/network/availability-recovery/src/tests.rs [hunk 3]
```diff
@@ -249,7 +287,6 @@ impl TestState {
 				let _ = tx.send(if with_data {
 					Some(self.available_data.clone())
 				} else {
-					println!("SENDING NONE");
```
Comment (eskimor) [inline node/network/availability-recovery/src/tests.rs:252]: why not use logging?

**Review by eskimor — APPROVED** [review]
Nice!

File: node/network/availability-recovery/src/tests.rs (hunk 3 above)
Comment (sandreim) [reply node/network/availability-recovery/src/tests.rs:252]: This seems to be a leftover, I'll fix it.

File: node/network/availability-recovery/src/lib.rs (hunk 2 above)
Comment (sandreim) [reply node/network/availability-recovery/src/lib.rs:894]: done

File: node/network/availability-recovery/src/lib.rs [hunk 4]
```diff
@@ -99,15 +99,49 @@ const TIMEOUT_START_NEW_REQUESTS: Duration = CHUNK_REQUEST_TIMEOUT;
...
+		}
+	}
+}
 /// The Availability Recovery Subsystem.
 pub struct AvailabilityRecoverySubsystem {
 	/// Do not request data from the availability store.
 	/// This is the useful for nodes where the
 	/// availability-store subsystem is not expected to run,
 	/// such as collators.
 	bypass_availability_store: bool,
```
Comment (sandreim) [inline node/network/availability-recovery/src/lib.rs:142]: Should remove this as we have None strategy.

File: node/network/availability-recovery/src/lib.rs [hunk 5]
```diff
@@ -99,15 +99,46 @@ const TIMEOUT_START_NEW_REQUESTS: Duration = CHUNK_REQUEST_TIMEOUT;
...
+	BackersFirstAlways,
+	/// We try the backing group first if PoV size is lower than specified, then fallback to validator chunks.
+	BackersFirstIfSizeLower(usize),
+	/// We always recover using validator chunks.
+	ChunksAlways,
+	/// No strategy, no work to be done.
+	/// This is the useful for nodes where the availability-store subsystem is not expected to run,
 	/// such as collators.
-	bypass_availability_store: bool,
+	None,
```
[the remark this answers is not in the document]
Comment (skunert) [reply node/network/availability-recovery/src/lib.rs:117]: This optimization is specifically for collators that are using the external relay chain RPC node. These nodes do not even run the av-store subsystem, so we do not request chunks.

[the remark this answers is not in the document]
Comment (skunert) [reply node/network/availability-recovery/src/lib.rs:117]: I think the naming of this `None` variant should be improved here. It implies that we do nothing, but we only skip asking the local store and recover directly from the network.

[the remark this answers is not in the document]
Comment (sandreim) [reply node/network/availability-recovery/src/lib.rs:117]: Oh, that makes sense. I'll switch to original name as variant.

[the remark this answers is not in the document]
Comment (sandreim) [reply node/network/availability-recovery/src/lib.rs]: fixed

**Review by ordian — APPROVED** [review]

**Comment by sandreim** [comment]
bot merge

**Pull request merged**